### PR TITLE
[WIP] IServiceProvider on runspace

### PR DESCRIPTION
**Pull request by powercode**
# PR Summary

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/System.Management.Automation/engine/SessionStatePublic.cs [hunk 1]
```diff
@@ -159,6 +160,15 @@ public PSModuleInfo Module
...
         }
 
+        /// <summary>
+        /// An <see cref="IServiceProvider"/> instance that provides custom services to other objects.
+        /// </summary>
+        public IServiceProvider ServiceProvider
+        {
+            get => _sessionState.ExecutionContext.ServiceProvider;
+            set => _sessionState.ExecutionContext.ServiceProvider = value;
+        }
```
Comment (SeeminglyScience) [inline src/System.Management.Automation/engine/SessionStatePublic.cs:170]: Having this API on `PSModuleInfo` makes it seem like any module can use this to set their own service provider. If two modules try to set up different service providers, then the last loaded will win and the other will break.

Have you considered using a `ConditionalWeakTable<PSModuleInfo, IServiceProvider>`?

Comment (powercode) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: Having lower-level modules be able to rebuild a service provider with additional services may be useful.

But perhaps something that should be made more specific?

The use case I want to cover is to be able to set an initial service provider from the InitialSessionState and make it possible to have modules recreate it with some added services.

That could perhaps be abstracted with a get only IServiceProvider property, and some
```csharp
IServiceProvider ServiceProvider { get;  }
void AddServices(IServiceCollection services)
```

But in that case, I think we would have to commit to some implementation of `IServiceProvider`.

Comment (SeeminglyScience) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: I still think it'll be potentially challenging for different module authors to use the same instance. Especially if they find themselves in a runspace with a different implementation.

What I was getting at with the weak table suggestion, you could have a common assembly with code like this:

```csharp
namespace MyOrg
{
    public static class Services
    {
        private static readonly ConditionalWeakTable<Runspace, IServiceProvider> s_services = new();

        public static IServiceProvider Get(Runspace runspace) => s_services.GetValue(runspace, Create);

        public static IServiceProvider GetFromTLS() => s_services.GetValue(Runspace.DefaultRunspace, Create);

        private static IServiceProvider Create(Runspace key)
        {
            // etc
        }
    }
}
```

Comment (powercode) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: That is quite similar to what I have been using (not with a ConditionalWeakTable, good suggestion!) 
but that doesn't really help with injecting something differently in tests, or am I missing something?

Comment (SeeminglyScience) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: Updated with how I'd handle a different service provider for tests. Could also just go with a static property that's like `ServiceProviderOverride`.

If that doesn't answer your question, can you provide an example of how you'd use it with the proposal for tests?

Comment (powercode) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: It is possible to work around this, as you pointed out, with some static map where I could register the runspace mapped to a service provider, but then either all modules would need such a mapping, or share some common library.

By adding this to the SessionState, it seems things become much simpler.

Comment (powercode) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: An alternative may be to have SessionState provide a PowerShell defined interface, where there is only a getter for the service provider, and some method like
```csharp
public interface IServices{
   IServiceProvider ServiceProvider {get;}
   void AddServicesAndRebuildProvider(IServiceCollection services);
}
```

Then there can only ever be a single ServiceProvider, but you can add services on a per-module basis.

Comment (SeeminglyScience) [reply src/System.Management.Automation/engine/SessionStatePublic.cs:169]: Yeah but I'm thinking if two modules are using a common library and they both register services of the same interface then that'll become an issue still

**Pull request closed without merging**